### PR TITLE
Dependency factory

### LINKED ISSUE #310: Dependency factory
Right now the logic to construct all the different components/dependencies is duplicated on almost every command. This makes kind of cumbersome to add more dependencies to any entity.

Add a package/method/entity to consolidate this logic in one place, making easier to add more entities and/or change an existing entity dependencies.

Context for components upgrades: this will facilitate a lot to inject more `executables` as dependencies on each eks-a component’s owner: like kubectl or helm (eg. Cilium). This is necessary if we want each owner to be able to follow its own upgrade process and just expose a common interface

### PR DESCRIPTION
**Pull request by g-gaston**
Resolves #310 

In order to facilitate adding/removing/changing the dependencies for our types and at the same time reduce the duplicated code/work at the `cmd` level (that's where we usually do DI), this OR add a factory to build a dependency graph.

It's a pretty simple implementation but it achieves:
* Only building whatever is necessary based on the input
* Only build each entity once
* Graph encapsulation (the caller doesn't need to know anything about the dependencies)
* Changing the graph is fairly simple

Some drawbacks:
* It requires to write a bit of boring code every time we want to add a new entity to the graph
* It requires to mark the dependencies for a certain entity explicitly (no inference)

There is definitely room for improvement but I believe the current state is good enough to start using it.

I also considered using https://github.com/google/wire instead, but with the amount of dependencies we have, it was going to take the same amount of time to learn how to use it than the time that saves by generating the boilerplate code. It's something that we could reconsider in the future if we start spending too much time adding entities to the factory.

There might be other places where we can use the factory (maybe E2E tests). I just started with the commands since they were the most obvious place.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/337#" title="Author self-approved">g-gaston</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [g-gaston]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by danbudris — APPROVED** [review]
so....to make sure I understand.

you're maintaining a stack of idempotent dependency creation functions; each ensures that its own dependencies are included in the stack before it, if necessary.

When we .Build() the dependency graph , each dependency is built in first-in-first-out order, ensuring that each dependency is built in the correct order, without repeating (the 'idempotent' nil checks).

Is that right?

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/337#pullrequestreview-765964739" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/337#" title="Author self-approved">g-gaston</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,g-gaston]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by g-gaston** [comment]
> so....to make sure I understand.
> 
> you're maintaining a stack of idempotent dependency creation functions; each ensures that its own dependencies are included in the stack before it, if necessary.
> 
> When we .Build() the dependency graph , each dependency is built in first-in-first-out order, ensuring that each dependency is built in the correct order, without repeating (the 'idempotent' nil checks).
> 
> Is that right?

That's correct :)

**Review by vivek-koppuru — COMMENTED** [review]
Neat! Looks a lot cleaner now 😄

File: pkg/dependencies/factory.go
Comment (g-gaston) [reply pkg/dependencies/factory.go:45]: Yeah, this method is just a helper when creating dependencies based on a particular `clusterSpec`. In that case, there are certain configurations that are always the same. The collector image is not always used but is always the same for the same `clusterSpec`, that's why is here.
To be precise, these `With*` methods in the helper are slightly different than the other ones, since they don't necessarily build dependencies, but they configure the factory to be able to build the real dependencies.

Regarding `New()`, I believe that's`NewFactory()`, isn't it? (it' called `NewFactory` instead of just `New` because it doesn't build the `dependencies` but a factory to build dependencies)

Comment (vivek-koppuru) [reply pkg/dependencies/factory.go:45]: I get the distinction now, this looks good to me

**Review by vivek-koppuru — APPROVED** [review]
/lgtm

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/337#pullrequestreview-765964739" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/337#" title="Author self-approved">g-gaston</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/337#pullrequestreview-767038874" title="LGTM">vivek-koppuru</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,g-gaston,vivek-koppuru]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/337#pullrequestreview-765964739" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/337#" title="Author self-approved">g-gaston</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/337#pullrequestreview-767066409" title="Approved">mrajashree</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/337#pullrequestreview-767038874" title="LGTM">vivek-koppuru</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,g-gaston,mrajashree,vivek-koppuru]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
